tasks/hyrax-metadata-updates/index.js

```diff
@@ -0,0 +1,318 @@
+  // we have to defend against that by reconstructing OnlineResources as
+  // a list from scratch
+  delete metadataCopy.Granule.OnlineResources;
+  metadataCopy.Granule.OnlineResources = {};
+  metadataCopy.Granule.OnlineResources.OnlineResource = [];
+
+  if (!isUndefined(metadata.Granule.OnlineResources)) {
+    const urls = metadata.Granule.OnlineResources.OnlineResource;
+    metadataCopy.Granule.OnlineResources.OnlineResource.push(urls);
+  }
+  const url = {
+    URL: hyraxUrl,
+    Description: 'OPeNDAP request URL',
+    Type: 'GET DATA : OPENDAP DATA'
+  };
+  metadataCopy.Granule.OnlineResources.OnlineResource.push(url);
+  return generateEcho10XMLString(metadataCopy.Granule);
+}
+
+/**
+ * addHyraxUrl
+ *
+ * @param {Object} metadata - the original metadata dom
+ * @param {boolean} isUmmG - UMM-G or ECHO10 metadata
+ * @param {string} hyraxUrl - the hyrax url
+ * @returns {string} - the updated metadata containing a Hyrax URL
+ */
+function addHyraxUrl(metadata, isUmmG, hyraxUrl) {
+  let updatedMetadata = null;
+  if (isUmmG === true) {
+    updatedMetadata = addHyraxUrlToUmmG(metadata, hyraxUrl);
+  } else {
+    updatedMetadata = addHyraxUrlToEcho10(metadata, hyraxUrl);
+  }
+  return updatedMetadata;
+}
+
+/**
+ * createDom
+ *
+ * @param {Object} metadataFileName file name
+ * @param {string} metadata - raw metadata
+ * @returns {Object} document object model and whether it is UMM-G
+ */
+async function createDom(metadataFileName, metadata) {
+  let isUmmG = false;
+  // Parse into DOM based on file extension
+  let dom = null;
+  if (isUMMGFile(metadataFileName)) {
+    dom = JSON.parse(metadata);
+    isUmmG = true;
+  } else if (isECHO10File(metadataFileName)) {
+    dom = await (promisify(xml2js.parseString))(metadata, xmlParseOptions);
+  } else {
+    throw new InvalidArgument(`Metadata file ${metadataFileName} is in unknown format`);
+  }
+  return { dom, isUmmG };
+}
+
+/**
+ * updateSingleGranule
+ *
+ * @param {Object} config
+ * @param {Object} granuleObject - granule files object
+ */
+async function updateSingleGranule(config, granuleObject) {
+  // Read in the metadata file
+  const metadataFile = granuleObject.files.find((f) => isCMRFilename(f.filename));
+  // If there is no metadata file, error out.
+  if (isUndefined(metadataFile)) {
+    throw new RecordDoesNotExist('There is no recogizable CMR metadata file in this granule object (*.cmr.xml or *.cmr.json)');
+  }
+  const { Bucket, Key } = parseS3Uri(metadataFile.filename);
+  const metadataResult = await getS3Object(Bucket, Key);
+
+  const tags = await s3GetObjectTagging(Bucket, Key);
+
+  // Extract the metadata file object
+  const metadata = metadataResult.Body.toString();
+  const { dom, isUmmG } = await createDom(metadataFile.name, metadata);
+  // Add OPeNDAP url
+  const hyraxUrl = await generateHyraxUrl(config, dom, isUmmG);
+  const updatedMetadata = await addHyraxUrl(dom, isUmmG, hyraxUrl);
+  // Validate updated metadata via CMR
+  if (isUmmG) {
```

Comment:
I started going down that route and then struggled with how things are now. I think it could be cleaned up but would need some refactoring done in cmr-client. Currently the echo10 validate takes an xml string and the umm g validation takes an object and didn't want to create a common utility function that would take a parameter that could be a string or an object depending on what the function will do. I don't want to go down a bigger refactor of all of that right now.